src/node/blockstorage.cpp

```diff
@@ -788,19 +788,24 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatFilePos& pos, c
     return true;
 }
 
-/** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
 FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, CChain& active_chain, const CChainParams& chainparams, const FlatFilePos* dbp)
 {
     unsigned int nBlockSize = ::GetSerializeSize(block, CLIENT_VERSION);
     FlatFilePos blockPos;
-    if (dbp != nullptr) {
+    const auto position_known {dbp != nullptr};
+    if (position_known) {
         blockPos = *dbp;
+    } else {
+        // when known, blockPos.nPos points at the offset of the block data in the blk file. that already accounts for
+        // the serialization header present in the file (the 4 magic message start bytes + the 4 length bytes = 8 bytes = BLOCK_SERIALIZATION_HEADER_SIZE).
+        // we add BLOCK_SERIALIZATION_HEADER_SIZE only for new blocks since they will have the serialization header added when written to disk.
+        nBlockSize += static_cast<unsigned int>(BLOCK_SERIALIZATION_HEADER_SIZE);
```

Comment:
I also think this comment would be helpful, except for the point below.

> ```c++
> // When position_known is true, WriteBlockToDisk will have already been
> // called, and it will have adjusted blockPos.nPos to point to the CBlock
> // data after the header
> ```
Isn't this incorrect? My understanding is that when `position_known` is true, `blockPos.nPos` was set by the `LoadExternalBlockFile` code which has read the header of an existing block file and move the pos beyond it - but not `WriteBlockToDisk`, which wasn't called for this block at all during this invocation of `bitcoind`.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24858#discussion_r913370952

> > ```c++
> > // When position_known is true, WriteBlockToDisk will have already been
> > // called, and it will have adjusted blockPos.nPos to point to the CBlock
> > // data after the header
> > ```
> 
> Isn't this incorrect?

I think it is correct 😎. The line that changes `blockPos.nPos` to point after the header instead of before the header is here:

https://github.com/bitcoin/bitcoin/blob/9fb2a2bc6768ab03fcada9155d52a16ce6f6a0cc/src/node/blockstorage.cpp#L676. 

This is the critical code calculating the block position which is saved in the `CBlockIndex` struct, and written to BlockIndexDb and used when `ReadBlockFromDisk` is called to read back block data.

> My understanding is that when `position_known` is true, `blockPos.nPos` was set by the `LoadExternalBlockFile` code which has read past the header in the existing block file and moved the position accordingly

I don't think `LoadExternalBlockFile` is relevant to normal operation, only used when importing or reindexing is requested.

> but not `WriteBlockToDisk`, which wasn't called for this block at all during this invocation of `bitcoind`.

It's true that `WriteBlockToDisk` may or may not have been called during this invocation of `bitcoind`, but it must have been called previously at some point, and the position it returns is written to BlockIndexDb and used to read blocks in future invocations.

Reply:
Hmm, but the suggested comment is not in `ReadBlockFromDisk` but in `SaveBlockToDisk`, and it is about the case when "position_known is true", or `dbp != nullptr`. I think this is only possible when this function is called via `AcceptBlock()` from `LoadExternalBlockFile` which has set the blockPos - the normal operation calls to `SaveBlockToDisk` (via `AcceptBlock()` and `ProcessNewBlock()`) have `dbp==nullptr`, so `position_known` is false and the other half of the comment applies.

Reply:
Sure, if you feel it is more useful for the comment to go into those things, feel free to make a suggestion along those lines. My suggestion was rejected anyway so maybe you will have better luck.

Also, I wrote this comment to describe a piece of code that I think is unnecessarily confusing. I would personally prefer not to have this piece of code, and not to have this comment explaining it, and instead use a solution like https://github.com/ryanofsky/bitcoin/commit/07c05a1b24bf42e57f8c229f505ed448a6df81f7 which does not require understanding external places where `SaveBlockToDisk` is called to understand, and only requires understanding internally how it always returns the same position.

Reply:
I don't have a strong preference - it seems that both approaches only affect the blockstorage-internal logic, and since neither solution changes the public interface `SaveBlockToDisk`, this sholdn't change what external places will need to know: In particular, that if they pass `dbp!=nullptr`, dbp needs to point right after the header and before the block data (which should be documented in `SaveBlockToDisk` imo).

However, I think that the internal function `FindBlockPos()` would be clearer if it was split up depending on whether the block is known, with so much code being conditional on it: If `dbp!=nullptr` , the function doesn't "find" anything new but only updates the block info statistics (and sometimes does a Flush which I'm not convinced is necessary). I had an initial go at this in https://github.com/mzumsande/bitcoin/commits/202207_refactor_findblockpos , which is outside the scope of this PR, but might make sense as a follow-up.

Reply:
I strongly agree with @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/24858/files#r895769516) (above) about consistency. I'd also note that adding 8 to `nBlockSize` is confusing because it's no longer a block size!